### DISCUSSION TITLE
Cluster replication over the internet

### DISCUSSION THREAD

**Discussion by michagruner** [seaweedfs/seaweedfs#1756] [Q&A]
Hi, seaweedfs seems a great thing. I'm currently using minio but intend to switch to seaweedfs because of the posix mounts which is not usable with minio in my setup using s3fs.

I have a question I'm not coming to an answer when digging through the docs: How may I realize a secure cluster replication over the internet? My scenario is as following: I would build a cluster at my home with redundancy. I have a reverse proxy with https to serve some of my services at home (currently gitea, s3). At my parents home I would like to install a raspberry with a huge hard disk attached that replicates changes over night and then shuts down. I would not like to expose anything at my parents home if that is possible. Could you please give me some rope here whether that is achiveable and if so, how?

thanks, Michael

**Comment by chrislusf** [comment]
The `weed filer.sync` process needs to access both clusters (filer and volume servers). You can put it in either cluster.
https://github.com/chrislusf/seaweedfs/wiki/Filer-Active-Active-cross-cluster-continuous-synchronization

To secure the access, https://github.com/chrislusf/seaweedfs/wiki/Security-Configuration

However, the network setup seems complicated. You will need to expose at least one cluster to the `weed filer.sync` process.

Maybe later need to add a `weed filer.proxy` to reverse proxy the requests.

**Comment by chrislusf** [comment]
I added a feature in version 2.22 so that `weed filer.sync` will only need to talk with the filers in each cluster.

So if this is the topo, and you want backup from clusterA to clusterB, 
```
clusterA <--> internet <--> clusterB
```
you can setup SSH tunnel from clusterB to the filer in clusterA on port 8888 and 18888. And then run `weed filer.sync` inside clusterB network. 
```
clusterA <--> internet <--> ssh tunnel to filer in clusterA => `weed filer.sync` => clusterB
```

**Reply by michagruner** [reply]
Wow, that is quick, thank you!

I did some research over the weekend. I think the best solution might be to use stunnel (http://stunnel.org) or ghostunnel on the internet facing end of each cluster. They provide mutual authentication, so the architecture could look like:

`filer A<-->stunnel-client<-->internet<-->stunnel-server<-->filer B`

**Reply by chrislusf** [reply]
This seems cleaner. Maybe share your experience later or write up in a wiki page? SeaweedFS wiki is open to everyone.

**Reply by michagruner** [reply]
Sorry, didn't see your ssh solution when writing my stunnel solution. Might work both. I will try and let you know, can't promise I'll be as quick with a solution as you were with yours, though. ;)

**Reply by dch** [reply]
I suggest using [spiped](https://www.tarsnap.com/spiped.html) e.g. https://blog.kempkens.io/posts/securing-http-traffic-with-spiped/ you will find it is both faster throughput, and generally more stable than using/needing ssh or stunnel. You'd still need to open the respective firewall ports.

It apparently has a go port now, https://github.com/dchest/gospiped maybe that would be a useful inbuilt wrapper between weedfs servers?

**Comment by dch** [comment]
the other thing that is of interest, is that you can run seaweedfs to connect itself over VPN IPs as well. My current setup uses https://zerotier.com/ to allow nodes to communicate, in my specific case, all weedfs servers are using IPv6 only but there's no particular requirement for that.

I would generally expect throughput to be lower than with spiped, but it is more than sufficient for our needs.
